### ISSUE TITLE
Computation expressions should support syntax desugaring for return!/yield! in tailcall positions

### ISSUE THREAD

**Issue by dsyme** [fsharp/fslang-suggestions#1006]
Computation expressions like task, seq, taskSeq, async, coroutine often need to support tailcalls in the computational structure, e.g.

```fsharp
let rec f n = 
    seq { 
        ...do something... ; 
        if n > 0 then
            yield! f (n-1) 
    }
```

Here tailcalling means that, once the first sequence is done its work, it delegates to the next sequence, and that chains of sequences either "cut out" or otherwise relinquish resources.  

Tailcall detection of `yield!` is built in to the F# compiler for its compilation of sequence expressions (see [this](https://github.com/dotnet/fsharp/blob/main/src/fsharp/FSharp.Core/seqcore.fs#L352)). However for other computation expressions this is harder. The problem is that `return!` and `yield!` are often used in non-tailcall positions, e.g. inside a try/with or try/finally, or sequencing of `yield!`.   This makes it awkward to implement tailcalling semantics, especially for computation expressions that can be compiled with resumable code.  Workarounds do exist but they increase the amount of code generated and have other problems.

Instead, we should simply augment computation expression de-sugaring to desugar `return!`/`yield!` to `ReturnFromFinal`/`YieldFromFinal` (or some other name) when they occur at the natural tailcall position if the method is present on the computation expression builder.  Likewise `do!` in the final position can translate to `ReturnFromFinal` if present.

**Comment by TysonMN** [comment]
This seems somewhat related to #721 to me.

**Comment by Lanayx** [comment]
Tasks tailcall support is much appreciated!

**Comment by AngelMunoz** [comment]
I'll come to add my two cents use case here :)

I'm writing an [async html rendering library](https://github.com/AngelMunoz/Hox) which can produce a single string via `StringBuilder` or the main use is to render the html in chunks via `IAsyncEnumerable<string>` so they can be streamed as is to the client from a server, this can be used to leverage browser's [incremental rendering](https://en.wikipedia.org/wiki/Incremental_rendering).

This could also be used in the future to provide other client side benefits such as out of order rendering and many many things.

I initially went for for a recursive approach (since HTML lives as a tree data structure) with IcedTasks' `ValueTask` support + `TaskSeq`, however I had to change my approach to a stack based approach instead with limited recursion, when depth is more than ~200ish nodes we'd rather await to convert the whole `IAsyncEnumerable<Node>` sequence to a list then process it otherwise we have crashes at runtime due stack overflows, this limits how often can chunks be emitted from our side.

If recursion was supported in `return!`/`yield!` it would allow to avoid buffering the async sequences before yielding the rendered content.

**Comment by brianrourkeboll** [comment]
@T-Gro should this be closed now that https://github.com/dotnet/fsharp/pull/18804 and https://github.com/fsharp/fslang-design/pull/809 are in F# 10?

**Comment by T-Gro** [comment]
Yes 👍 .

**Comment by Lanayx** [comment]
@T-Gro Do we have another ticket for adding tailcall support for task builder, or one should be created?

**Comment by brianrourkeboll** [comment]
> [@T-Gro](https://github.com/T-Gro) Do we have another ticket for adding tailcall support for task builder, or one should be created?

I don't think there is one, so you could open one.

But @majocha did [note](https://github.com/dotnet/fsharp/pull/18804#issuecomment-3144108179):

> OTOH Applying this to TaskBuilder is very tricky and probably hopeless, given its type signature.
> It returns just a running task and binds to awaitables that also may or may not be already running. I cannot come up with anything that wouldn't be horrendous (like for example communicating between TaskBuilder state machines via an AsyncLocal side channel to signal that we're executing in a recursive taill-call context).

😅😅

**Comment by Lanayx** [comment]
I see, also wonder if runtime async will change the perspective.

**Comment by majocha** [comment]
The issue is very practical:

Given our current implementation in tasks.fs

```fsharp
        member inline this.ReturnFrom(task: Task<'T>) : TaskCode<'T, 'T> =
            this.Bind(task, this.Return)
```

The idea of tail call optimization here would be to abandon the intermediate tasks and continue just the next one in the chain. We could save CPU by not doing all the SetResult on intermediate tasks and save memory (and stack) by not keeping all of them while the computation runs.

But in the code above there is no way to tell where the bound `task` comes from or who else awaits it, so we can't just drop it.

This problem does not exist in cold multi-start asyncs where notions of computation and its invocation are not conflated.
